Auto tests for PR: kubevirt/kubevirt#1839

sHaggYcaT opened the pull request:
**What this PR does / why we need it**:

Auto tests for PR:  https://github.com/kubevirt/kubevirt/pull/1839

**Release note:**
```release-note
NONE
```

MarSik commented:
@sHaggYcaT I do not think you have to check all the VM creation, just use those as functions. I am pretty sure there was a way to call a method and create VM from yaml without having to implement everything on you own.

See for example https://github.com/kubevirt/kubevirt-ansible/blob/master/tests/high_performance_vm_test.go#L46

Currently it is very hard to see the test code in all the boilerplate around it.

sHaggYcaT commented:
@MarSik ^^updated version ^^
Do not ready for merge, but almost done

MarSik commented on tests/sockets_cores_threads_test.go:
```diff
@@ -0,0 +1,484 @@
+	return arguments
+}
+
+var _ = Describe("numaSupport", func() {
```
NUMA? This has nothing to do with numa.. guest topology might be better name here. Also shouldn't you annotate this with the test plan ID somehow so the results are automatically connected with it?

sHaggYcaT replied:
>NUMA
Just old variable, I'll rename it. I've already rename everything about NUMA, but in this place forget to do it

>Also shouldn't you annotate this with the test plan ID somehow so the results are automatically connected with it?

I'm going to do this work it in another PR for all my test plans together. Now I just want to merge this code ASAP  (before it outdated, and before it requires rebase/difficult merges, etc)

ILpinto replied:
Please fix

sHaggYcaT replied:
It was already fixed

MarSik commented on tests/sockets_cores_threads_test.go:
```diff
@@ -0,0 +1,484 @@
+	"time"
+)
+
+type Config struct {
```
Why not reuse the kubevirt/kubernetes API objects?

sHaggYcaT replied:
Could you please show an example?

sHaggYcaT replied:
^^ @MarSik ^^

MarSik replied:
https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3144 this structure implements the same I think

sHaggYcaT replied:
I've just did it

sHaggYcaT commented:
>make sure you properly cleanup the started VMs
I've tested it in the openshift local installation and in official CI. If test passed, then everything delete. If not passed, failed VMI not delete (and it's helpful for debug). It's not problem for github CI, because it creates dedicated environment for each patch. And it's helpful if you run test manually using our internal CI

sHaggYcaT commented:
@MarSik how to remove "WIP" flag?

lukas-bednar commented:
@sHaggYcaT could you please try to add following value to LagoInitFile.yml.j2

```
diff --git a/playbooks/provider/lago/LagoInitFile.yml.j2 b/playbooks/provider/lago/LagoInitFile.yml.j2
index d6fb07ac..8cbcd7a4 100644
--- a/playbooks/provider/lago/LagoInitFile.yml.j2
+++ b/playbooks/provider/lago/LagoInitFile.yml.j2
@@ -6,6 +6,7 @@ nat-settings: &nat-settings
     management: False
 
 vm-common-settings: &vm-common-settings
+    vcpu: 4
     root-password: 123456
     service_provider: systemd
     artifacts:
```

sHaggYcaT commented:
> @sHaggYcaT could you please try to add following value to LagoInitFile.yml.j2

Thank you @lukas-bednar

lukas-bednar commented on playbooks/provider/lago/LagoInitFile.yml.j2:
```diff
@@ -6,6 +6,7 @@ nat-settings: &nat-settings
     management: False
 
 vm-common-settings: &vm-common-settings
+    vCpu: 4
```
it is lowercased `vcpu`, I am afraid that it might cause that this option is going to ignored :-O

sHaggYcaT replied:
@lukas-bednar looks like it doesn't work :(

sHaggYcaT replied:
Forget to say, your latest suggestion works

ILpinto commented on tests/framework/util.go:
```diff
@@ -134,3 +134,55 @@ func RemoveDataVolume(dvName string, namespace string) {
 	Expect(err).ToNot(HaveOccurred())
 }
+
+func CreateVMorVMIfromManifest(manifest, nameSpace string) string {
```
See tests/framework/types.go
Use: func (vm VirtualMachine) Create() (string, string, error)

sHaggYcaT replied:
Done

ILpinto commented on tests/framework/util.go:
```diff
@@ -134,3 +134,55 @@ func RemoveDataVolume(dvName string, namespace string) {
+
+// Find out how many VMs with specific resource needs can be run in the cluster
+
+func min(a, b int) int {
```
use https://golang.org/pkg/math/#Min

sHaggYcaT replied:
OK, thank you

sHaggYcaT replied:
Done

ILpinto commented on tests/sockets_cores_threads_test.go:
```diff
@@ -0,0 +1,484 @@
+	virtClient, err := kubecli.GetKubevirtClient()
+	ktests.PanicOnError(err)
+	address_common := "tests/manifests/sockets_cores_threads/"
+	const TestSystemNamespace = "kubevirt-test-default"
```
import
	tests "kubevirt.io/kubevirt-ansible/tests/framework"
use like:
tests.NamespaceTestDefault
remove this const

sHaggYcaT replied:
Done

ILpinto commented on tests/sockets_cores_threads_test.go:
```diff
@@ -0,0 +1,484 @@
+ *
+ */
+
+package tests_test
```
Add polarion test ids to all test cases

sHaggYcaT replied:
@ILpinto  done

lukas-bednar commented on tests/framework/types.go:
```diff
@@ -39,22 +40,22 @@ type VirtualMachine struct {
 func (vm VirtualMachine) Create() (string, string, error) {
 	args := []string{"create", "-f", vm.Manifest}
-	return ktests.RunCommandWithNS(vm.Namespace, ktests.KubeVirtOcPath, args...)
+	return ktests.RunCommandWithNS(vm.Namespace, ocName, args...)
```
`ktests.KubeVirtOcPath` is configurable from command line, I don't find this change as an improvement.

sHaggYcaT replied:
@lukas-bednar I've rewrote it using hack/run-tests.sh (as you suggested in the email) and shell ENV

sHaggYcaT commented:
Other tests fails, every test carried by this PR, passed CI(please check log).
@MarSik @ILpinto could you please approve CI? Then, we'll ask to merge it.

sHaggYcaT commented:
/hold cancel

sHaggYcaT commented:
/do-not-merge cancel

MarSik commented on tests/sockets_cores_threads_test.go:
```diff
@@ -0,0 +1,493 @@
+						//prevent server overloading - CI workaround (i/o error)
+						time.Sleep(15 * time.Second)
+
+						go func(sockets int, cores int, threads int, wg *sync.WaitGroup) {
```
Are you really running all the combinations in paralel? And why don't you use the global sockets, cores, threads and wg? Are the arguments to this method really necessary? The code is slighly unreadable because of them and the necessary parentheses block at line 485.

Can some golang expect check this please?

sHaggYcaT replied:
>Are you really running all the combinations in paralel?

I don't understand what happen, but if I don't use timeout, the code panics. It runs about 4-5 VMs in one time by fact.

sHaggYcaT replied:
I will do additional research: decrease timeout. And temporary delete network-related tests from PR

vladikr replied:
perhaps you want to move this function definition elsewhere and just call it here?
Martin is right, it's harder to read this way.

You can just define this in the top of the test `runMyVM := func(sockets int, cores int, threads int, wg*sync.WaitGroup) {`
Then call it here go runMyVM(sockets, cores, threads, &wg)

sHaggYcaT replied:
Vadik, thank you. I've Implemented it.

ILpinto commented on tests/framework/types.go:
```diff
@@ -39,22 +44,22 @@ type VirtualMachine struct {
 func (vm VirtualMachine) Start() (string, string, error) {
 	args := []string{"start", vm.Name}
-	return ktests.RunCommandWithNS(vm.Namespace, ktests.KubeVirtVirtctlPath, args...)
+	return ktests.RunCommandWithNS(vm.Namespace, ocName, args...)
```
oc don't start VM, virtctl is.
and  like Lukas comment please use the path exist already.

ILpinto commented on tests/framework/types.go:
```diff
@@ -39,22 +44,22 @@ type VirtualMachine struct {
 func (vm VirtualMachine) Stop() (string, string, error) {
 	args := []string{"stop", vm.Name}
-	return ktests.RunCommandWithNS(vm.Namespace, ktests.KubeVirtVirtctlPath, args...)
+	return ktests.RunCommandWithNS(vm.Namespace, ocName, args...)
```
so use virtctl

sHaggYcaT replied:
fixed

ILpinto commented on tests/framework/types.go:
```diff
@@ -39,22 +44,22 @@ type VirtualMachine struct {
 func (vm VirtualMachine) Delete() (string, string, error) {
 	args := []string{"delete", vm.Type, vm.Name}
-	return ktests.RunCommandWithNS(vm.Namespace, ktests.KubeVirtVirtctlPath, args...)
+	return ktests.RunCommandWithNS(vm.Namespace, ocName, args...)
```
same use path

ILpinto commented on tests/framework/types.go:
```diff
@@ -73,7 +78,7 @@ func (vm VirtualMachine) IsRunning() (bool, error) {
 func (vm VirtualMachine) GetVMInfo(spec string) (string, string, error) {
 	args := []string{"get", vm.Type, vm.Name, "--template", spec}
-	return ktests.RunCommandWithNS(vm.Namespace, ktests.KubeVirtOcPath, args...)
+	return ktests.RunCommandWithNS(vm.Namespace, ocName, args...)
```
same use path

ILpinto commented on tests/framework/types.go:
```diff
@@ -100,7 +105,7 @@ func (vm VirtualMachine) ProcessTemplate() (string, error) {
 	args = append(args, vm.TemplateParams...)
 
-	output, cmderr, err := ktests.RunCommandWithNS(NamespaceTestTemplate, ktests.KubeVirtOcPath, args...)
+	output, cmderr, err := ktests.RunCommandWithNS(NamespaceTestTemplate, ocName, args...)
```
same use path

ILpinto commented on tests/manifests/sockets_cores_threads/vm-template-fedora-only-threads.yaml:
```diff
@@ -0,0 +1,67 @@
+        terminationGracePeriodSeconds: 0
+        volumes:
+        - containerDisk:
+            image: kubevirt/cirros-registry-disk-demo
```
Fedora or Cirros?

ILpinto commented on tests/manifests/sockets_cores_threads/vm-template-fedora-only-sockets.yaml:
```diff
@@ -0,0 +1,67 @@
+        terminationGracePeriodSeconds: 0
+        volumes:
+        - containerDisk:
+            image: kubevirt/cirros-registry-disk-demo
```
Fedora or Cirros?

ILpinto commented on tests/manifests/sockets_cores_threads/vm-template-fedora-only-sockets-and-threads.yaml:
```diff
@@ -0,0 +1,71 @@
+        terminationGracePeriodSeconds: 0
+        volumes:
+        - containerDisk:
+            image: kubevirt/cirros-registry-disk-demo
```
Fedora or Cirros?

ILpinto commented on tests/manifests/sockets_cores_threads/vm-template-fedora-only-sockets-and-cores.yaml:
```diff
@@ -0,0 +1,73 @@
+        terminationGracePeriodSeconds: 0
+        volumes:
+        - containerDisk:
+            image: kubevirt/cirros-registry-disk-demo
```
Fedora or Cirros?

sHaggYcaT replied:
fixed

ILpinto commented on tests/manifests/sockets_cores_threads/vm-template-fedora-only-cores.yaml:
```diff
@@ -0,0 +1,67 @@
+        terminationGracePeriodSeconds: 0
+        volumes:
+        - containerDisk:
+            image: kubevirt/cirros-registry-disk-demo
```
Fedora or Cirros?

sHaggYcaT replied:
Forget to comment - fixed few days ago

ILpinto commented on tests/manifests/sockets_cores_threads/vm-template-fedora-only-cores-and-threads.yaml:
```diff
@@ -0,0 +1,71 @@
+        terminationGracePeriodSeconds: 0
+        volumes:
+        - containerDisk:
+            image: kubevirt/cirros-registry-disk-demo
```
Fedora or Cirros?

sHaggYcaT replied:
fixed

ILpinto commented on tests/manifests/sockets_cores_threads/vm-template-fedora-no-sockets-cores-and-threads.yaml:
```diff
@@ -0,0 +1,62 @@
+        terminationGracePeriodSeconds: 0
+        volumes:
+        - containerDisk:
+            image: kubevirt/cirros-registry-disk-demo
```
Fedora or Cirros?

sHaggYcaT replied:
fixed

vladikr commented on tests/sockets_cores_threads_test.go:
```diff
@@ -0,0 +1,493 @@
+							}, 60*time.Second)
+							Expect(err).ToNot(HaveOccurred(), "Should report number of sockets")
+
+							/*By("2.2 Checking the number of cores in guest OS")
```
It's better to remove comments

sHaggYcaT replied:
Thank you. Just forget to uncomment it

vladikr commented on tests/sockets_cores_threads_test.go:
```diff
@@ -0,0 +1,493 @@
+
+			for sockets := 0; sockets < 3; sockets++ {
+				for cores := 0; cores < 3; cores++ {
+					for threads := 0; threads < 3; threads++ {
```
I guess, it's better to start from 1?, I see 0 is never used or am I wrong?

sHaggYcaT replied:
No, it uses. If zero, then code uses other template

vladikr commented:
Just for the record, I'm in favor of merging this PR. I wasn't clear if this repo is alive or not.

However, looking at the test results, I see that this test is failing on `Insufficient memory` and I fear that running so many tests in parallel maybe an issue. At least we need to provide a timeout waiting for a virtual machine to actually start. Take a look at kubevirt tests [2], it provides such functionality. 
I think that panics you've earlier observed were also caused by insufficient resources rather than synchronization.

Overall, not necessarily related to this test, when dealing with getting results from async tasks, it's better to use gomega functions, such as `Eventually` [3] - there are multiple examples in kubevirt tests. 



[1]
```
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2019-04-25T06:16:45Z"
    message: '0/3 nodes are available: 1 Insufficient devices.kubevirt.io/kvm, 1 Insufficient
      devices.kubevirt.io/tun, 1 Insufficient devices.kubevirt.io/vhost-net, 1 node(s)
      didn''t match node selector, 2 Insufficient memory.'
    reason: Unschedulable
    status: "False"
    type: PodScheduled
  phase: Scheduling
```

[2] https://github.com/kubevirt/kubevirt/blob/master/tests/utils.go#L1183
[3] https://onsi.github.io/gomega/#eventually

sHaggYcaT commented:
@vladikr thank you. I've fixed goroutine bug using the following approach:

1) launching vms in the main thread
2) So, next vm can launch only when previous launched (and kubernetes returns VMI's object)
3) Virtual machines boot in async mode

Also I've did several other minor improvements, and fixed one missed (because CI works time to time) bug.

sHaggYcaT commented:
Failed in webUI deploy:

17:03:26  TASK [kubevirt_web_ui : Wait until Web UI is ready: 'v1.4.0-13'] ***************

sHaggYcaT commented:
ci please

sHaggYcaT commented:
ci test please

sHaggYcaT commented:
@vladikr thank you.
@MarSik permitted few work-arounds  to fix CI. So, I not sure, maybe my fixes broken something. In this case let me know please. 
my tests finished without errors: check  [rfe_id:1443]. So, I'll merge this PR
